### PR TITLE
Output tensor specified via out= must be on the same device as inputs for dot & vdot

### LINKED ISSUE #55561: A RuntimeError should be raised when input and out= tensor are on different devices for torch.{dot, vdot}
Reference: https://github.com/pytorch/pytorch/pull/55492
Failure from the OpInfo generated test for `torch.{dot, vdot}`
`RuntimeError not raised : Expected RuntimeError when calling with input.device=cuda:0 and out.device=cpu`

### PR DESCRIPTION
**Pull request by imaginary-person**
Fixes #55561.

1. Added checks to ensure that Output tensor specified via out= must be on the same device as inputs for `dot` & `vdot`.
2. Unskipped `test_out` for `dot` & `vdot`. 
3. Also changed the `tensordot` implementation to check if both input tensors are on the same device as the output tensor.

### CODE REVIEW

**Comment by anjali411** [comment]
FYI https://github.com/pytorch/pytorch/pull/56409 has landed

**Comment by imaginary-person** [comment]
Thanks for the info, @anjali411! 
I've already imported that PR & skipped `test_out` for both `dot` & `vdot`. CI checks are currently running.

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (imaginary-person) [inline aten/src/ATen/native/LinearAlgebra.cpp:1230]: @anjali411, here, I'm only checking if `self` and `result` are on the same device, but not `tensor`.
Please advise if I should add the following line as well:
`checkSameDevice("dot_out", result, tensor);`

There's a function `checkAllSameGPU()` for checking if any number of tensors are on the same GPU,
but no such function exists for all devices (such as `checkAllSameDevice`).

BTW, the CI failure is unrelated.
`test_out` passed for both `dot` & `vdot`.

**Comment by imaginary-person** [comment]
Hello @anjali411, I'm now checking if both the input tensors are on the same device as the output tensor, and revised the error message to display `dot` instead of `dot_out`, for example.
Thanks!

File: aten/src/ATen/native/Linear.cpp
Comment (imaginary-person) [inline aten/src/ATen/native/Linear.cpp:653]: I had submitted a PR earlier that was checking the device of `input1`, but not `input2`. 😞

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (imaginary-person) [reply aten/src/ATen/native/LinearAlgebra.cpp]: Looks like using `checkSameDevice("dot", result, tensor);` results in a weird build failure

Comment (imaginary-person) [reply aten/src/ATen/native/LinearAlgebra.cpp]:
```
/var/lib/jenkins/workspace/aten/src/ATen/native/LinearAlgebra.cpp:1240:3: error: no matching function for call to 'checkSameDevice'
 checkSameDevice("vdot", result, tensor);
 ../../aten/src/ATen/native/LinearAlgebraUtils.h:485:20: note: candidate function not viable: no overload of 'tensor' matching 'at::Tensor' for 3rd argument
```

File: aten/src/ATen/native/Linear.cpp
Comment (imaginary-person) [inline aten/src/ATen/native/Linear.cpp:656]: Hello @heitorschueroff, in the tensordot PR a few days back, I should've checked the device type of both the input tensors, and not just the first one.

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (imaginary-person) [reply aten/src/ATen/native/LinearAlgebra.cpp]: I simply renamed `tensor` to `other`

File: aten/src/ATen/native/Linear.cpp
Comment (mruberry) [reply aten/src/ATen/native/Linear.cpp:656]: Sounds good; thanks for noticing

Comment (mruberry) [reply aten/src/ATen/native/Linear.cpp:656]: The code here is a little confusing because the C++ calls the inputs `input1` and `input2`, but the user-facing names from the documentation are `a` and `b`, and error messages should reflect the user's experience and not the code; so this is a great error message but it needs a slight tweak to replace `input1` with `a` and `input2` with `b`

File: aten/src/ATen/native/LinearAlgebra.cpp
Comment (mruberry) [inline aten/src/ATen/native/LinearAlgebra.cpp:1234]: would you also fix this error message and replace "self" with "input", here?

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:3180]: Awesome!

**Comment by mruberry** [comment]
Hey @imaginary-person! This is really cool; I made a couple minor comments about reflecting the docs in the error messages, but otherwise this looks great

Also it looks like not all the CI builds triggered, so maybe pushing the docs update will trigger them and give us a signal?

**Comment by imaginary-person** [comment]
All CI checks had passed earlier, so I had been using `[skip ci]` to make minor changes, such as changing comments or renaming.
However, I'll trigger all builds this time around. Thanks!

**Comment by mruberry** [comment]
> All CI checks had passed earlier, so I had been using `[skip ci]` to make minor changes, such as changing comments or renaming.
> However, I'll trigger all builds this time around. Thanks!

I appreciate your wanting to respect the ci, but thanks for triggering the builds. It just makes me feel better to sanity-check the PR CI before merging

File: aten/src/ATen/native/Linear.cpp
Comment (imaginary-person) [reply aten/src/ATen/native/Linear.cpp]: It seems that even [`checkSameDevice`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebraUtils.h#L485-L491) in `LinearAlgebraUtils.h` is also a bit lacking in terms of UX.
It only accepts one input, so if `checkSameDevice` is called multiple times for multiple input tensors, then it may not be immediately clear to a user as to which input is not on the same device as the output tensor. So, if you'd like, I can use a `TORCH_CHECK` for `dot` & `vdot` in this PR instead.

`checkAllSameGPU` in `TensorUtils.h` is better in the sense that it lets one provide individual names for tensors, which are included in error messages. But as the name suggests, it only checks if any given number of tensors are on the same GPU.

I proposed adding `checkAllSameDevice` in PR #56628 to check if multiple tensors are on the same device, but then a simple `TORCH_CHECK` might be more appropriate.

Comment (mruberry) [reply aten/src/ATen/native/Linear.cpp]: Using simpler existing utilities, like TORCH_CHECK, sounds like a good solution for now

I agree that eventually we should review and see if there's a simpler, more universal solution

**Review by mruberry — APPROVED** [review]
Nice work, @imaginary-person. Cool to have these fixed.

**Pull request closed without merging**
